cEP-0026.md

```diff
@@ -0,0 +1,189 @@
+| Version  | 1.0                                       |
+| Title    | Optimize caching for the NextGen-Core     |
+| Authors  | Palash Nigam <mailto:npalash25@gmail.com> |
+| Status   | Implementation Due                        |
+| Type     | Feature                                   |
+
+## Abstract
+
+This document describes how a caching mechanism (an enhanced version of the one
+being used by the old core) will be implemented and integrated with the
+NextGen-Core as a part of this
+[GSoC project](https://summerofcode.withgoogle.com/projects/#6434190552203264).
+
+## Introduction
+
+Since the NextGen-Core aims to provide faster speed and better performance for
+coala-runs on large codebases it needs caching to achieve this kind of
+performance.
+
+The current caching used by the old core can be improved upon and integrated
+with the NextGen-Core by adding a few more enhancements to the current caching
+mechanism like `FileFactory` class to load files, `Directory` class
+to interface with the directories, ignoring unmodified directories during
+successive runs and adding cache control flags for better usability and
+providing the user various options to perform caching according to a project's
+needs.
+
+## Implementation
+
+The implementation will be divided into four parts
+
+- Prototype for `FileFactory` class implementation.
+- Prototype for `Directory` class implementation.
+- Ignore directories functionality.
+- Cache control flags.
+
+### Prototype of FileFactory class implementation
+
+This class will be implemented according to the Factory design pattern and will
+act as a proxy to dealing with files. This class will have various methods to
+provide the file contents in different forms for e.g. string, binary
+format, etc. The biggest benefit that we will get out of this class would be
+that we can replace the actual file contents in our file-dict with these
+objects which will itself be a performance improvement.
+
+The following is the prototype for `FileFactory` class.
+
+```python
+import linecache
+import os
+import pickle
+
+from coalib.core.PersistentHash import persistent_hash
+
+
+class FileFactory:
+    def __init__(self, filename, parent):
+        self.filename = filename # absolute file path
+        self.parent = os.path.abspath(os.path.join(p, os.pardir)) # parent dir
+        self._content = open(self.filename)
+
+    def open(filename, newline=None):
+        return FileFactory(filename, newline=newline)
+
+    def timestamp(self):
+        # get the last modified timestamp of the file
+        return os.path.gettime(self.path)
+
+    def file_as_string(self, encoding=None):
+        return "".join(self._content)
+
+    def file_as_bytes(self):
+        # return the pickled version of the self._content
+        pickle.dumps(self.filename)
+
+    def hash(self):
+        # hash the absolute file path to get a unique hash
+        return persistent_hash(self.filename)
```

Comment:
If you want to support persistent hashing, you should implement the pickling methods. But for the prototype it's currently okay :+1:

Reply:
(then doing `persistent_hash(file_factory)` would work too)

Reply:
Exactly, and when you override the pickling interface, you can control persistent-hash behavior :) then you can control which data influences the final hash value :)